api/output_policy.go

```diff
@@ -0,0 +1,197 @@
+package api
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+
+	retryablehttp "github.com/hashicorp/go-retryablehttp"
+	"github.com/hashicorp/vault/sdk/helper/jsonutil"
+)
+
+const (
+	ErrOutputPolicyRequest = "output policy request"
+)
+
+var LastOutputPolicyError *OutputPolicyError
+
+type OutputPolicyError struct {
+	*retryablehttp.Request
+	VaultClient     *Client
+	parsingError    error
+	parsedHCLString string
+}
+
+func (d *OutputPolicyError) Error() string {
+	if d.parsedHCLString == "" {
+		d.parseRequest()
+		if d.parsingError != nil {
+			return d.parsingError.Error()
+		}
+	}
+
+	return ErrOutputPolicyRequest
+}
+
+// Builds a sample policy document from the request
+func (d *OutputPolicyError) parseRequest() {
+	capabilities := []string{}
+	switch d.Request.Method {
+	case "GET":
+		capabilities = append(capabilities, "read")
+	case "LIST":
+		capabilities = append(capabilities, "list")
+	case "POST", "PUT":
+		capabilities = append(capabilities, "create")
+		capabilities = append(capabilities, "update")
+	case "PATCH":
+		capabilities = append(capabilities, "patch")
+	case "DELETE":
+		capabilities = append(capabilities, "delete")
+	}
+
+	// trim the Vault address and v1 from the front of the path
+	url := d.Request.URL.String()
+	apiAddrPrefix := fmt.Sprintf("%sv1/", d.VaultClient.config.Address)
+	path := strings.Trim(url, apiAddrPrefix)
+
+	// determine whether to add sudo capability
+	needsSudo, err := isSudoPath(d.VaultClient, path)
+	if err != nil {
+		d.parsingError = err
+		return
+	}
+	if needsSudo {
+		capabilities = append(capabilities, "sudo")
+	}
+
+	capStr := strings.Join(capabilities, `", "`)
+	d.parsedHCLString = fmt.Sprintf(
```

Comment:
Thiank you, this ended up being the comment that finally got me to refactor the whole pattern a bit (for both output flags)! Hooray. 

Just in changing it to return the string I found that storing the error on the object seemed unnecessary as well, so now hopefully the logic is a little simpler as it doesn't need to check Error() in generatePolicy(), it can just check HCLString() and see if that returns an error.